### PR TITLE
SI-8032 Provides an implicit macro for materialization of Liftable for case classes

### PR DESCRIPTION
**Pull request by folone**
Now that development of 2.12 has started, reopening this PR.

Case classes can now be used in quasiquotes.

``` scala
Welcome to Scala version 2.11.1-20140505-184318-2e6e13ff25 (Java HotSpot(TM) 64-Bit Server VM, Java 1.7.0_51).
Type in expressions to have them evaluated.
Type :help for more information.

scala> import scala.reflect.runtime.{universe => u}
import scala.reflect.runtime.{universe=>u}

scala>  import u._
import u._

scala> case class Test(str: String, i: Int, lst: List[Double])
defined class Test

scala> val test = Test("test", 1, List(2.0, 3.5))
test: Test = Test(test,1,List(2.0, 3.5))

scala> showRaw(q"$test")
res0: String = Apply(Ident(Test), List(Literal(Constant("test")), Literal(Constant(1)), Apply(Select(Select(Select(Ident(scala), TermName("collection")), TermName("immutable")), TermName("List")), List(Literal(Constant(2.0)), Literal(Constant(3.5))))))

scala> val cm = runtimeMirror(getClass.getClassLoader)
cm: reflect.runtime.universe.Mirror = JavaMirror with scala.tools.nsc.interpreter.IMain$TranslatingClassLoader@75293656 of type class scala.tools.nsc.interpreter.IMain$TranslatingClassLoader with classpath [(memory)] and parent being scala.reflect.internal.util.ScalaClassLoader$URLClassLoader@64ae8431 of type class scala.reflect.internal.util.ScalaClassLoader$URLClassLoader with classpath [file:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/resources.jar,file:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/rt.jar,file:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/jsse.jar,file:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/jce.jar,file:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Con...
scala> import scala.tools.reflect.ToolBox
import scala.tools.reflect.ToolBox

scala> val tb = cm.mkToolBox()
tb: scala.tools.reflect.ToolBox[reflect.runtime.universe.type] = scala.tools.reflect.ToolBoxFactory$ToolBoxImpl@641c66ff

scala> tb.eval(q"$test")
res1: Any = Test(test,1,List(2.0, 3.5))

scala> .asInstanceOf[Test]
res2: Test = Test(test,1,List(2.0, 3.5))
```

Review by @xeno-by, @densh, and @retronym please.

### CODE REVIEW

**Comment by folone** [comment]
PR-validation job does not seem to kick in. Is this the right branch to send PRs for 2.12, or should I still send them to master?

**Comment by xeno-by** [comment]
Why not 2.11.x btw?

**Comment by folone** [comment]
@xeno-by Quoting @retronym from discussion in #3326:

> I'm still not comfortable adding this macro for 2.11, as I think we should work on the foundations of quasiquotes before we add this convenience. Closing, but please resubmit after 2.12 dev starts.

Maybe backport it later?

**Comment by retronym** [comment]
I'm still working on the infrastructure for the 2.12.x branch (PR validation, nightlies, bumping version.properties etc). I just edited the GitHub milestone `2.12.1-M1` to add the magic comment "Merge to 2.12.x" and labelled with issue with the milestone. That seems to have picked up the build.

**Comment by densh** [comment]
I've did some [derivative work](https://github.com/densh/core/commit/21be7ffa163d8fafb921c0e449fbbe6cf1ee0e19#diff-0e35640dae65413a046bbfd488421107R11) out of this to materialize transitive liftables for Palladium trees. Unlike this pr that macro eagerly expands all transitive dependencies. This adds some boilerplate but prevents unnecessary re-expansion of macro at every use site. I think we need to discuss this a bit more before merging into 2.12.

File: src/reflect/scala/reflect/api/StandardLiftables.scala [hunk 1]
```diff
@@ -233,3 +237,98 @@ trait StandardLiftables { self: Universe =>
...
+     * q"Apply($constructor, List(..$arguments))"
+     */
+    val reflect = Apply(Select(prefix, TermName("Apply")),
+                        List(constructor,
+                          Apply(Ident(definitions.ListModule),
+                        arguments)))
+    val implicitName = TermName(symbol.name.encoded ++ "Liftable")
+    /* Tree produced by the following quasiquote:
+     * q"""
+     *   implicit object $implicitName extends Liftable[$T] {
```
Comment (retronym) [inline src/reflect/scala/reflect/api/StandardLiftables.scala:306]: Why implicit?

File: src/reflect/scala/reflect/api/StandardLiftables.scala [hunk 2]
```diff
@@ -233,3 +237,98 @@ trait StandardLiftables { self: Universe =>
...
+  /* All the trees (except one that has a special comment) in this macro are hygienic
+   * (and there is a test to check that). That is, all TermNames used
+   * (specifically: Tree, Apply, Liftable, List, implicitly, reify) are fully qualified,
+   * and will not bind to terms with same names, declared in the expansion context.
+   * Each tree is commented with a simplified, unhygienic quasiquote, that explains
+   * what that tree is supposed to do.
+   */
+  def impl[T: c.WeakTypeTag](c: Context): c.Tree = {
+    import c.universe._
+    val Select(prefix, _) = c.prefix.tree
```
Comment (retronym) [inline src/reflect/scala/reflect/api/StandardLiftables.scala:253]: Why can you assume this tree shape for `c.prefix`?

**Pull request closed without merging**